private/buf/bufgen/features.go

```diff
@@ -22,86 +22,189 @@ import (
+			// Plugin supports editions, and files include editions. So make sure
+			// the plugin supports precisely the right editions.
+			requiredEditions := make([]descriptorpb.Edition, 0, len(required.editions))
+			for edition := range required.editions {
+				requiredEditions = append(requiredEditions, edition)
+			}
+			sort.Slice(requiredEditions, func(i, j int) bool {
+				return requiredEditions[i] < requiredEditions[j]
+			})
+			for _, requiredEdition := range requiredEditions {
+				if int32(requiredEdition) < response.GetMinimumEdition() ||
+					int32(requiredEdition) > response.GetMaximumEdition() {
+					failed.editions[requiredEdition] = required.editions[requiredEdition]
+					failedEditions = append(failedEditions, requiredEdition)
+				}
+			}
+		}
+
+		pluginName := configs[responseIndex].Name()
+		if len(failedFeatures) > 0 {
+			_, _ = fmt.Fprintf(
+				container.Stderr(),
+				"Plugin %q does not support required feature(s).\n",
+				pluginName)
 			sort.Slice(failedFeatures, func(i, j int) bool {
-				return failedFeatures[i].Number() < failedFeatures[j].Number()
+				return failedFeatures[i] < failedFeatures[j]
 			})
 			for _, feature := range failedFeatures {
-				files := failed[feature]
-				_, _ = fmt.Fprintf(container.Stderr(), "  Feature %q is required by %d file(s):\n",
+				files := failed.flags[feature]
+				_, _ = fmt.Fprintf(
+					container.Stderr(),
+					"  Feature %q is required by %d file(s):\n",
 					featureName(feature), len(files))
-				_, _ = fmt.Fprintf(container.Stderr(), "    %s\n", strings.Join(files, ","))
+				_, _ = fmt.Fprintf(
+					container.Stderr(),
+					"    %s\n",
+					strings.Join(files, ","))
+			}
+		}
+
+		if len(failedEditions) > 0 {
+			_, _ = fmt.Fprintf(
+				container.Stderr(),
+				"Plugin %q does not support required edition(s).\n",
+				pluginName)
+			sort.Slice(failedEditions, func(i, j int) bool {
+				return failedEditions[i] < failedEditions[j]
+			})
+			for _, edition := range failedEditions {
+				files := failed.editions[edition]
+				_, _ = fmt.Fprintf(
+					container.Stderr(),
+					"  Edition %q is required by %d file(s):\n",
+					editionName(edition), len(files))
+				_, _ = fmt.Fprintf(
+					container.Stderr(),
+					"    %s\n",
+					strings.Join(files, ","))
 			}
 		}
+
+		if len(failedFeatures) > 0 || len(failedEditions) > 0 {
+			failedPlugins = append(failedPlugins, pluginName)
+		}
+	}
+	switch len(failedPlugins) {
+	case 0:
+		return nil
+	case 1:
+		return fmt.Errorf("plugin %s is unable to generate code for all input files", failedPlugins[0])
+	default:
+		return fmt.Errorf("plugins [%v] are unable to generate code for all input files", strings.Join(failedPlugins, ","))
 	}
 }
 
 func featureName(feature pluginpb.CodeGeneratorResponse_Feature) string {
 	// FEATURE_PROTO3_OPTIONAL -> "proto3 optional"
+	return enumReadableName(feature, "FEATURE")
+}
+
+func editionName(edition descriptorpb.Edition) string {
```

Comment:
What's the purpose of this?

Reply:
It gets a human-readable name for the edition, to show in error messages. Was that unclear? Maybe I don't understand the question.

Reply:
I would expect/think it's more on point to print out EDITION_2023 instead of printing our own/custom name for the edition, I wouldn't strip the prefix